components/layout_thread_2020/lib.rs

```diff
@@ -1258,3 +1271,46 @@ impl RegisteredSpeculativePainters for RegisteredPaintersImpl {
             .map(|painter| painter as &dyn RegisteredSpeculativePainter)
     }
 }
+
+#[derive(Debug)]
+struct LayoutFontMetricsProvider(Arc<ReentrantMutex<FontCacheThread>>);
+
+impl FontMetricsProvider for LayoutFontMetricsProvider {
+    fn query_font_metrics(
+        &self,
+        _vertical: bool,
+        font: &Font,
+        base_size: CSSPixelLength,
+        _in_media_query: bool,
+        _retrieve_math_scales: bool,
+    ) -> FontMetrics {
+        layout::context::with_thread_local_font_context(&self.0, move |font_context| {
+            let Some(servo_metrics) = font_context
+                .font_group_with_size(ServoArc::new(font.clone()), base_size.px())
+                .borrow_mut()
+                .first(font_context)
+                .map(|font| font.borrow().metrics.clone())
+            else {
+                return Default::default();
+            };
+
+            // Only use the x height of this font if it is non-zero. Some fonts return
+            // inaccurate metrics, which shouldn't be used.
+            let x_height = if !servo_metrics.x_height.is_zero() {
+                Some(CSSPixelLength::new(servo_metrics.x_height.to_f32_px()))
+            } else {
+                None
+            };
```

Comment:
```suggestion
            let x_height = Some(servo_metrics.x_height)
                .filter(|x_height| !x_height.is_zero())
                .map(CSSPixelLength::from);
```

Reply:
There's definitely a tradeoff here between being terse and readable, but I think it isn't a big deal. I've made this change.